plugin/evm/vm.go

```diff
@@ -301,21 +299,29 @@ func (vm *VM) Initialize(
 		header.Extra = append(header.Extra, hid...)
 	})
 	chain.SetOnFinalizeAndAssemble(func(state *state.StateDB, txs []*types.Transaction) ([]byte, error) {
-		select {
-		case atx := <-vm.pendingAtomicTxs:
-			if err := atx.UnsignedTx.(UnsignedAtomicTx).EVMStateTransfer(vm, state); err != nil {
-				vm.newBlockChan <- nil
+		if tx, exists := vm.mempool.NextTx(); exists {
+			if err := tx.UnsignedTx.(UnsignedAtomicTx).EVMStateTransfer(vm, state); err != nil {
+				log.Error("Atomic transaction failed verification", "txID", tx.ID(), "error", err)
+				// Discard the transaction from the mempool on failed verification.
```

Comment:
is there a reason we don't call `vm.newBlockChan <- nil` in these error cases anymore?

Reply:
Yes, if the atomic transaction fails verification, we should still try to build a block consisting of only normal transactions if we can.

Reply:
Good catch